File: app/src/main/java/com/mohsenoid/rickandmorty/injection/DependenciesProvider.java
package com.mohsenoid.rickandmorty.injection;

import android.app.Application;

import com.mohsenoid.rickandmorty.data.RepositoryImpl;
import com.mohsenoid.rickandmorty.data.db.Db;
import com.mohsenoid.rickandmorty.data.db.DbImpl;
import com.mohsenoid.rickandmorty.data.mapper.CharacterMapper;
import com.mohsenoid.rickandmorty.data.mapper.EpisodeMapper;
import com.mohsenoid.rickandmorty.data.mapper.LocationMapper;
import com.mohsenoid.rickandmorty.data.mapper.OriginMapper;
import com.mohsenoid.rickandmorty.data.network.NetworkClient;
import com.mohsenoid.rickandmorty.data.network.NetworkClientImpl;
import com.mohsenoid.rickandmorty.data.network.NetworkConstants;
import com.mohsenoid.rickandmorty.data.network.NetworkHelper;
import com.mohsenoid.rickandmorty.data.network.NetworkHelperImpl;
import com.mohsenoid.rickandmorty.domain.Repository;
import com.mohsenoid.rickandmorty.util.config.ConfigProvider;
import com.mohsenoid.rickandmorty.util.config.ConfigProviderImpl;
import com.mohsenoid.rickandmorty.util.executor.IoTaskExecutor;
import com.mohsenoid.rickandmorty.util.executor.MainTaskExecutor;
import com.mohsenoid.rickandmorty.util.executor.TaskExecutor;
import com.mohsenoid.rickandmorty.view.character.details.CharacterDetailsContract;
import com.mohsenoid.rickandmorty.view.character.details.CharacterDetailsFragment;
import com.mohsenoid.rickandmorty.view.character.details.CharacterDetailsPresenter;
import com.mohsenoid.rickandmorty.view.character.list.CharacterListContract;
import com.mohsenoid.rickandmorty.view.character.list.CharacterListFragment;
import com.mohsenoid.rickandmorty.view.character.list.CharacterListPresenter;
import com.mohsenoid.rickandmorty.view.character.list.adapter.CharacterListAdapter;
import com.mohsenoid.rickandmorty.view.episode.list.EpisodeListContract;
import com.mohsenoid.rickandmorty.view.episode.list.EpisodeListFragment;
import com.mohsenoid.rickandmorty.view.episode.list.EpisodeListPresenter;
import com.mohsenoid.rickandmorty.view.episode.list.adapter.EpisodeListAdapter;
import com.mohsenoid.rickandmorty.view.util.ImageDownloader;
import com.mohsenoid.rickandmorty.view.util.ImageDownloaderImpl;

import java.util.List;

public class DependenciesProvider {

    private final Application context;

    public DependenciesProvider(Application context) {
        this.context = context;
    }

    private Db getDatastore() {
        return DbImpl.getInstance(context);
    }

    private NetworkHelper getNetworkHelper() {
        return NetworkHelperImpl.getInstance(NetworkConstants.BASE_URL);
    }

    private NetworkClient getApiClient() {
        NetworkHelper networkHelper = getNetworkHelper();
        return NetworkClientImpl.getInstance(networkHelper);
    }

    private TaskExecutor getIoTaskExecutor() {
        return IoTaskExecutor.getInstance();
    }

    private TaskExecutor getMainTaskExecutor() {
        return MainTaskExecutor.getInstance();
    }

    private ConfigProvider getConfigProvider() {
        return ConfigProviderImpl.getInstance(context);
    }

    private EpisodeMapper getEpisodeMapper() {
        return EpisodeMapper.getInstance();
    }

    private OriginMapper getOriginMapper() {
        return OriginMapper.getInstance();
    }

    private LocationMapper getLocationMapper() {
        return LocationMapper.getInstance();
    }

    private CharacterMapper getCharacterMapper() {
        OriginMapper originMapper = getOriginMapper();
        LocationMapper locationMapper = getLocationMapper();

        return CharacterMapper.getInstance(originMapper, locationMapper);
    }

    private Repository getRepository() {
        Db db = getDatastore();
        NetworkClient networkClient = getApiClient();
        TaskExecutor ioTaskExecutor = getIoTaskExecutor();
        TaskExecutor mainTaskExecutor = getMainTaskExecutor();
        ConfigProvider configProvider = getConfigProvider();
        EpisodeMapper episodeMapper = getEpisodeMapper();
        CharacterMapper characterMapper = getCharacterMapper();

        return RepositoryImpl.getInstance(db, networkClient, ioTaskExecutor, mainTaskExecutor, configProvider, episodeMapper, characterMapper);
    }

    private String getCacheDirectoryPath() {
        return context.getCacheDir().getAbsolutePath();
    }

    public ImageDownloader getImageDownloader() {
        NetworkHelper networkHelper = getNetworkHelper();
        String cacheDirectoryPath = getCacheDirectoryPath();
        TaskExecutor ioTaskExecutor = getIoTaskExecutor();
        TaskExecutor mainTaskExecutor = getMainTaskExecutor();

        return ImageDownloaderImpl.getInstance(networkHelper, cacheDirectoryPath, ioTaskExecutor, mainTaskExecutor);
    }

    public EpisodeListFragment getEpisodeListFragment() {
        return EpisodeListFragment.newInstance();
    }

    public EpisodeListContract.Presenter getEpisodesListFragmentPresenter() {
        Repository repository = getRepository();
        ConfigProvider configProvider = getConfigProvider();

        return new EpisodeListPresenter(repository, configProvider);
    }

    public EpisodeListAdapter getEpisodesListAdapter(EpisodeListAdapter.ClickListener listener) {
        return new EpisodeListAdapter(listener);
    }

    public CharacterListFragment getCharacterListFragment(List<Integer> characterIds) {
        return CharacterListFragment.newInstance(characterIds);
    }

    public CharacterListContract.Presenter getCharacterListFragmentPresenter() {
        Repository repository = getRepository();
        ConfigProvider configProvider = getConfigProvider();

        return new CharacterListPresenter(repository, configProvider);
    }

    public CharacterListAdapter getCharacterListAdapter(CharacterListAdapter.ClickListener listener) {
        ImageDownloader imageDownloader = getImageDownloader();
        return new CharacterListAdapter(imageDownloader, listener);
    }

    public CharacterDetailsFragment getCharacterDetailsFragment(int characterId) {
        return CharacterDetailsFragment.newInstance(characterId);
    }

    public CharacterDetailsContract.Presenter getCharacterDetailsFragmentPresenter() {
        Repository repository = getRepository();
        ConfigProvider configProvider = getConfigProvider();

        return new CharacterDetailsPresenter(repository, configProvider);
    }
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/injection/DependenciesProvider.kt
package com.mohsenoid.rickandmorty.injection

import android.app.Application
import androidx.room.Room
import com.jakewharton.retrofit2.converter.kotlinx.serialization.asConverterFactory
import com.mohsenoid.rickandmorty.BuildConfig
import com.mohsenoid.rickandmorty.data.RepositoryImpl
import com.mohsenoid.rickandmorty.data.db.Db
import com.mohsenoid.rickandmorty.data.db.DbConstants
import com.mohsenoid.rickandmorty.data.db.dto.DbCharacterModel
import com.mohsenoid.rickandmorty.data.db.dto.DbEpisodeModel
import com.mohsenoid.rickandmorty.data.db.dto.DbLocationModel
import com.mohsenoid.rickandmorty.data.db.dto.DbOriginModel
import com.mohsenoid.rickandmorty.data.mapper.CharacterDbMapper
import com.mohsenoid.rickandmorty.data.mapper.CharacterEntityMapper
import com.mohsenoid.rickandmorty.data.mapper.EpisodeDbMapper
import com.mohsenoid.rickandmorty.data.mapper.EpisodeEntityMapper
import com.mohsenoid.rickandmorty.data.mapper.LocationDbMapper
import com.mohsenoid.rickandmorty.data.mapper.LocationEntityMapper
import com.mohsenoid.rickandmorty.data.mapper.Mapper
import com.mohsenoid.rickandmorty.data.mapper.OriginDbMapper
import com.mohsenoid.rickandmorty.data.mapper.OriginEntityMapper
import com.mohsenoid.rickandmorty.data.network.NetworkClient
import com.mohsenoid.rickandmorty.data.network.NetworkConstants
import com.mohsenoid.rickandmorty.data.network.dto.NetworkCharacterModel
import com.mohsenoid.rickandmorty.data.network.dto.NetworkEpisodeModel
import com.mohsenoid.rickandmorty.data.network.dto.NetworkLocationModel
import com.mohsenoid.rickandmorty.data.network.dto.NetworkOriginModel
import com.mohsenoid.rickandmorty.domain.Repository
import com.mohsenoid.rickandmorty.domain.entity.CharacterEntity
import com.mohsenoid.rickandmorty.domain.entity.EpisodeEntity
import com.mohsenoid.rickandmorty.domain.entity.LocationEntity
import com.mohsenoid.rickandmorty.domain.entity.OriginEntity
import com.mohsenoid.rickandmorty.util.config.ConfigProvider
import com.mohsenoid.rickandmorty.util.config.ConfigProviderImpl
import com.mohsenoid.rickandmorty.util.dispatcher.AppDispatcherProvider
import com.mohsenoid.rickandmorty.util.dispatcher.DispatcherProvider
import com.mohsenoid.rickandmorty.view.character.details.CharacterDetailsContract
import com.mohsenoid.rickandmorty.view.character.details.CharacterDetailsFragment
import com.mohsenoid.rickandmorty.view.character.details.CharacterDetailsPresenter
import com.mohsenoid.rickandmorty.view.character.list.CharacterListContract
import com.mohsenoid.rickandmorty.view.character.list.CharacterListFragment
import com.mohsenoid.rickandmorty.view.character.list.CharacterListPresenter
import com.mohsenoid.rickandmorty.view.character.list.adapter.CharacterListAdapter
import com.mohsenoid.rickandmorty.view.episode.list.EpisodeListContract
import com.mohsenoid.rickandmorty.view.episode.list.EpisodeListFragment
import com.mohsenoid.rickandmorty.view.episode.list.EpisodeListPresenter
import com.mohsenoid.rickandmorty.view.episode.list.adapter.EpisodeListAdapter
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonConfiguration
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Converter
import retrofit2.Retrofit
import java.net.UnknownHostException

class DependenciesProvider(private val context: Application) {

    private val db: Db by lazy {
        Room.databaseBuilder(context, Db::class.java, DbConstants.DATABASE_NAME)
            .fallbackToDestructiveMigration()
            .build()
    }

    private val baseUrl: HttpUrl by lazy {
        NetworkConstants.BASE_URL.toHttpUrlOrNull()
            ?: throw UnknownHostException("Invalid host: " + NetworkConstants.BASE_URL)
    }

    private val json: Json by lazy {
        Json(JsonConfiguration.Stable)
    }

    private val jsonConverterFactory: Converter.Factory by lazy {
        val contentType = "application/json".toMediaType()
        json.asConverterFactory(contentType)
    }

    private val loggingInterceptor: HttpLoggingInterceptor by lazy {
        HttpLoggingInterceptor().apply {
            level = HttpLoggingInterceptor.Level.BODY
        }
    }

    private val okHttpClient: OkHttpClient by lazy {
        OkHttpClient.Builder().apply {
            if (BuildConfig.DEBUG)
                addInterceptor(loggingInterceptor)
        }
            .build()
    }

    private val retrofit: Retrofit by lazy {

        Retrofit.Builder()
            .baseUrl(baseUrl)
            .addConverterFactory(jsonConverterFactory)
            .client(okHttpClient)
            .build()
    }

    private val networkClient: NetworkClient by lazy {
        retrofit.create(NetworkClient::class.java)
    }

    val dispatcherProvider: DispatcherProvider by lazy {
        AppDispatcherProvider()
    }

    private val configProvider: ConfigProvider by lazy {
        ConfigProviderImpl(context)
    }

    private val episodeDbMapper: Mapper<NetworkEpisodeModel, DbEpisodeModel> by lazy {
        EpisodeDbMapper()
    }

    private val episodeEntityMapper: Mapper<DbEpisodeModel, EpisodeEntity> by lazy {
        EpisodeEntityMapper()
    }

    private val originDbMapper: Mapper<NetworkOriginModel, DbOriginModel> by lazy {
        OriginDbMapper()
    }

    private val originEntityMapper: Mapper<DbOriginModel, OriginEntity> by lazy {
        OriginEntityMapper()
    }

    private val locationDbMapper: Mapper<NetworkLocationModel, DbLocationModel> by lazy {
        LocationDbMapper()
    }

    private val locationEntityMapper: Mapper<DbLocationModel, LocationEntity> by lazy {
        LocationEntityMapper()
    }

    private val characterDbMapper: Mapper<NetworkCharacterModel, DbCharacterModel> by lazy {
        CharacterDbMapper(originDbMapper, locationDbMapper)
    }

    private val characterEntityMapper: Mapper<DbCharacterModel, CharacterEntity> by lazy {
        CharacterEntityMapper(originEntityMapper, locationEntityMapper)
    }

    private val repository: Repository by lazy {
        RepositoryImpl(
            characterDao = db.characterDao,
            episodeDao = db.episodeDao,
            networkClient = networkClient,
            dispatcherProvider = dispatcherProvider,
            configProvider = configProvider,
            episodeDbMapper = episodeDbMapper,
            episodeEntityMapper = episodeEntityMapper,
            characterDbMapper = characterDbMapper,
            characterEntityMapper = characterEntityMapper
        )
    }

    fun getEpisodeListFragment(): EpisodeListFragment {
        return EpisodeListFragment.newInstance()
    }

    fun getEpisodesListFragmentPresenter(): EpisodeListContract.Presenter {
        return EpisodeListPresenter(repository, configProvider)
    }

    fun getEpisodesListAdapter(listener: EpisodeListAdapter.ClickListener): EpisodeListAdapter {
        return EpisodeListAdapter(listener)
    }

    fun getCharacterListFragment(characterIds: List<Int>): CharacterListFragment {
        return CharacterListFragment.newInstance(characterIds)
    }

    fun getCharacterListFragmentPresenter(): CharacterListContract.Presenter {
        return CharacterListPresenter(repository, configProvider)
    }

    fun getCharacterListAdapter(listener: CharacterListAdapter.ClickListener): CharacterListAdapter {
        return CharacterListAdapter(dispatcherProvider, listener)
    }

    fun getCharacterDetailsFragment(characterId: Int): CharacterDetailsFragment {
        return CharacterDetailsFragment.newInstance(characterId)
    }

    fun getCharacterDetailsFragmentPresenter(): CharacterDetailsContract.Presenter {
        return CharacterDetailsPresenter(repository, configProvider)
    }
}
